### PR TITLE
feat: support discovery of symlinked modules

### LINKED ISSUE #1611: Symlinks in Terragrunt
Hi!

I have found that Terragrunt does not work with symlinks as expected. Or expected but just in a different way from Terraform.
For example, lets assume that we have such folders structure:
```
/aaa/bbb/ccc/cloudfront/development
/aaa/bbb/ccc/cloudfront/development/stage
/aaa/bbb/ccc/modules/cloudfront
~/cloudfront_development -> /aaa/bbb/ccc/cloudfront/development
```
Now, when I am trying to run `~/cloudfront_development$ terragrunt run-all plan` it fails with an error:
```
ERRO[0000] Could not find any subfolders with Terragrunt configuration files
ERRO[0000] Unable to determine underlying exit code, so Terragrunt will exit with error code 1
```
But commands `~/cloudfront_development$ terragrunt plan` and `~/cloudfront_development/stage$ terragrunt plan` separatelly works fine.

When I have navigated to `/aaa/bbb/ccc/cloudfront/development` directly and then run `$ terragrunt run-all plan` then it worked.

Also, I was trying to get rid of Terraform modules declaration and to declare scripts sources in the Terragrunt configuration `terraform` block:
replace modules.tf
```
module "stage" {
  source                          = "../../../modules/cloudfront"
}
```
with terragrunt.hcl
```
terraform {
  source = "../../..//modules/cloudfront"
}
```
But `~/cloudfront_development/stage$ terragrunt plan` has also exited with error:
```
ERRO[0004] Hit multiple errors:
stat /home/modules: no such file or directory
ERRO[0004] Unable to determine underlying exit code, so Terragrunt will exit with error code 1
```
As you can see, the Terraform is recognizing the source path correctly even when running its commands from the symlink location.

So is it an expected behavior with symlinks in Terragrunt?

### PR DESCRIPTION
**Pull request by aslafy-z**
## Description

Support discovery of symlinked modules with the help of https://github.com/edwardrf/symwalk.

Fixes #1611 

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [ ] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes

Added support for discovery of symlinked modules

### CODE REVIEW

**Comment by aslafy-z** [comment]
@denis256 This PR is ready for review, let me know if it need changes.

**Comment by denis256** [comment]
Conflicting files
go.mod

**Comment by aslafy-z** [comment]
@denis256 I just rebased!

File: config/config.go
Comment (denis256) [inline config/config.go:609]: I was wondering why `filepath.Walk()` wasn't replaced in other places, for example in `terraform/source.go`, I think it may lead to the failure of detection of changes in modules

https://github.com/search?q=repo%3Agruntwork-io%2Fterragrunt%20filepath.Walk&type=code

Comment (aslafy-z) [reply config/config.go:609]: I made the change.

**Comment by denis256** [comment]
Circle ci build is failing:
```
--> darwin/arm64 error: exit status 1
Stderr: # github.com/gruntwork-io/terragrunt/terraform
terraform/source.go:60:10: undefined: symwalk
```

**Comment by aslafy-z** [comment]
@denis256 I added the missing import and rebased the PR.

**Review by levkohimins — APPROVED** [review]
LGTM!

**Comment by aslafy-z** [comment]
@denis256 quick ping, can you take a look?
@levkohimins any chance you assign some other maintainer so this PR goes through? 
Thanks! 🙏

**Comment by anton-bogdanov** [comment]
i'm very interested in this PR getting merged too! Thanks

**Comment by levkohimins** [comment]
> @denis256 quick ping, can you take a look? @levkohimins any chance you assign some other maintainer so this PR goes through? Thanks! 🙏

Hi @aslafy-z, thanks for the PR, could you please resolve the conflicting files?

**Comment by aslafy-z** [comment]
@levkohimins I just resolved conflicts.

**Comment by denis256** [comment]
It is interesting that the same test fails multiple times:

image

**Comment by aslafy-z** [comment]
@levkohimins @denis256 Can you please trigger tests again?

**Comment by denis256** [comment]
It looks like tests still fail on `TestInitSkipCache`
[TestInitSkipCache.txt](https://github.com/user-attachments/files/16496223/TestInitSkipCache.txt)

**Comment by denis256** [comment]
Example of local failing:

[failing-test-3101.txt](https://github.com/user-attachments/files/16496442/failing-test-3101.txt)

On master branch test is passing, after switching on this PR branch, it is failing

**Comment by aslafy-z** [comment]
I just pushed [`dd6b989` (#3101)](https://github.com/gruntwork-io/terragrunt/pull/3101/commits/dd6b989e5f67f394a85f82c63e194fd3aa1be084) that fixes the failing test TestInitSkipCache by reverting the use of symwalk. For some reason, Source.EncodeSourceVersion was returning a different version on every CLI run (consecutive function calls returns the same value).

**Comment by denis256** [comment]
Triggered another run

**Comment by anton-bogdanov** [comment]
Hi @denis256 @levkohimins any chance you could review and merge the PR, since the tests are passing now? Thanks!
cc @aslafy-z

**Comment by denis256** [comment]
Before we merge this,I think we need to take a closer look at the implications of `.terragrunt-cache` invalidation, particularly concerning the modified time of symlinks. 
It's important that we carefully test any edge cases to ensure that these changes don't introduce unexpected behavior.

**Comment by yhakbar** [comment]
Hey @aslafy-z , this PR has gone stale. 

Would you mind rebasing, then following @denis256 's recommendations regarding introducing more testing?

**Comment by yhakbar** [comment]
This is still stale. I'm going to close it out. 

If you would like to open it again, please make sure to rebase the pull request, lint and test it locally as appropriate, then request review.

**Pull request closed without merging**
